homeassistant/components/modbus/__init__.py

```diff
@@ -213,6 +213,12 @@ def read_coils(self, unit, address, count):
             kwargs = {"unit": unit} if unit else {}
             return self._client.read_coils(address, count, **kwargs)
 
+    def read_discrete_inputs(self, unit, address, count):
+        """Read discrete inputs."""
+        with self._lock:
```

Comment:
We normally don't want sync locks in our thread pool. We shouldn't address this in this PR, but it would be something to improve for the future.

It would be ok to replace the sync lock with an asyncio lock. That would require changing the context around the affected methods from sync to async.

Reply:
Maybe I will take a look at it someday. I try to rework Modbus one step at a time. It's not the most perfect integration, but it gets better.